Fixes and improvements to pm policy

siddharthgpta opened the pull request:
This PR addresses 2 issues:
1. Buggy implementation of the latency requests that result in incorrect max latencies allowed by the system
2. Fix implementation of the next event registration. Since the next event is not registered with the kernel timeout manager, the sleep will be incorrect.

siddharthgpta committed:
pm: policy: fix latency request updates

The latency requests made to the power management subsystem is currently
buggy. The main issue is that the latency values are treated as int32_ts
rather than uint32_ts. By casting the latency request as an int32_t the
max latency request gets stuck at -1.

Consider the following scenario:
1. Thread A registers request with latency of 2ms
2. Thread B registers request with latency of SYS_FOREVER_US

In this scenario:
- The snippet runs for the request from thread A, new_max_latency_us is set
  to 2000.
- The snippet runs for the request from thread B, new_max_latency_us is
  2000 however the request from B is SYS_FOREVER_US(-1). Since -1 < 2000,
  new_max_latency_us is now set to SYS_FOREVER_US(-1).
- At the end of update_max_latency(), max_latency_cyc is set to -1.

This addresses this problem by converting the types of the latency
variables to uint32_t which should be the case to begin with since latency
requests can only be positive values.

Signed-off-by: Siddharth Gupta <siddharth.gupta2694@gmail.com>

kartben commented on include/zephyr/pm/policy.h:
```diff
@@ -58,6 +58,23 @@ struct pm_policy_latency_request {
+ */
+typedef void (*pm_policy_next_event_cb_t)(int64_t next_event);
+
+/**
+ * @brief Next event subscription.
+ *
+ * @note All fields in this structure are meant for private usage.
+ */
+struct pm_policy_event_subscription {
+	sys_snode_t node;
+	pm_policy_next_event_cb_t cb;
+};
```
```suggestion
/**
 * @brief Next event subscription.
 */
struct pm_policy_event_subscription {
	/**
	 * @cond INTERNAL_HIDDEN
	 */
	sys_snode_t node;
	pm_policy_next_event_cb_t cb;
	/**
	 * @endcond
	 */
};
```

kartben replied:
@siddharthgpta not sure if this is an oversight, but fyi your most recent push doesn't seem to have addressed this (in fact, it's not clear if it contains any change at all :) )

siddharthgpta replied:
Yes looks like I messed up the branch I pushed to. Thanks for flagging it again, I'll fix it!

ceolin reviewed (changes requested):
```
The latency requests made to the power management subsystem is currently
buggy. The main issue is that the latency values are treated as int32_ts
rather than uint32_ts. By casting the latency request as an int32_t the
max latency request gets stuck at -1 because new_max_latency_us is
initialized to SYS_FOREVER_US which is defined as -1. Any requests that
come in which are lesser than INT32_MAX will be ignored since they are not
less than -1.
```

I can't see how this happens because of the first condition in the branch:
```
		if ((new_max_latency_us == SYS_FOREVER_US) ||
		    ((int32_t)req->value_us < new_max_latency_us)) {
			new_max_latency_us = (int32_t)req->value_us;
		}
```
in the very first iteration `new_max_latency_us` will get assigned the contents of the first element.

The problem though is that we can have any latency in between [0, UINT32_MAX] but since we are internally representing it as `int32_t` we are losing precision.

Changing the type seems ok (and correct) but the problem description is different unless I have missed something.

Have you reproduced the problem you described ? If so, please ignore my comment and add test case please.

The new APIs are equivalent to what we have for latency so it is ok IMHO. Can you add a testcase for them in the policy_api test ?

siddharthgpta commented:
I'll update the message and add more information about the error that happens!

siddharthgpta commented:
While going through the policy_api tests I realized that I don't understand the usage of the latency APIs. So I have an example that should help clarify my question, and potentially might help close this PR without any changes.

Is the following usage correct?
```
static struct pm_policy_latency_request my_dummy_req;

static int my_dummy_thread(void *p1, void *p2, void *p3)
{
    struct my_dummy_msg msg;

    while (true) {
        k_msgq_get(&my_dummy_msgq, &msg, K_FOREVER);
        pm_policy_latency_request_update(&my_dummy_req, 2000);
        ...
        /* Do work that requires latency < 2ms */
        ...
        pm_policy_latency_request_update(&my_dummy_req, SYS_FOREVER_US);
    }
}

static int dummy_driver_init(const struct device *dev)
{
    k_thread_create(
        ...,
        my_dummy_thread,
        ...
    );

    /**
     * Add a latency request to be used by the thread
     */
    pm_policy_latency_request_add(&my_dummy_req, SYS_FOREVER_US);
}
```

Or do we want to remove the latency request each time till the next time we get a message? Seems inefficient to remove and add the latency from the list each time imho. However if the API is written that way, I guess I have no option but to conform to the design.

The problem I am trying to fix happens when I do `pm_policy_latency_request_update(&req, SYS_FOREVER_US)` and some other thread wants lower latency.

siddharthgpta commented:
To add more information about the error for context. Following is the error that I observe.

Focusing on this snippet only:
```
		if ((new_max_latency_us == SYS_FOREVER_US) ||
		    ((int32_t)req->value_us < new_max_latency_us)) {
			new_max_latency_us = (int32_t)req->value_us;
		}
```

Thread A registers request with latency of 2ms
Thread B registers request with latency of SYS_FOREVER_US

1. The snippet runs for the request from thread A, `new_max_latency_us` is set to 2000.
2. The snippet runs for the request from thread B, `new_max_latency_us` is 2000 however the request from B is `SYS_FOREVER_US`(-1). Since -1 < 2000, `new_max_latency_us` is now set to `SYS_FOREVER_US`(-1).
3. At the end of `update_max_latency()` max_latency_cyc is set to -1

Now when `pm_system_suspend()` runs, `pm_policy_next_state()` it will think there are no latency requests, and put the device in the lowest power mode possible depending on the ticks it can sleep for.

siddharthgpta commented:
@ceolin, do you have any idea? If not, could you tag someone who would know what the expected usage is?

ceolin commented on include/zephyr/pm/policy.h:
```diff
@@ -278,6 +311,18 @@ static inline void pm_policy_event_unregister(struct pm_policy_event *evt)
 	ARG_UNUSED(evt);
 }
+
+void pm_policy_next_event_subscribe(struct pm_policy_event_subscription *req,
```
`static inline` otherwise each object will have its own implementation of this function and the linker will complain

ceolin commented on include/zephyr/pm/policy.h:
```diff
@@ -278,6 +311,18 @@ static inline void pm_policy_event_unregister(struct pm_policy_event *evt)
+	ARG_UNUSED(cb);
+}
+
+void pm_policy_next_event_unsubscribe(struct pm_policy_event_subscription *req)
```
ditto

ceolin commented on include/zephyr/pm/policy.h:
```diff
@@ -58,6 +58,23 @@ struct pm_policy_latency_request {
+ *
+ * @param next_event Next event in cycles.
+ */
+typedef void (*pm_policy_next_event_cb_t)(int64_t next_event);
```
Please can you address this comment ?

ceolin commented:
> To add more information about the error, following is what I observe.
> 
> Focusing on this snippet:
> 
> ```
> 		if ((new_max_latency_us == SYS_FOREVER_US) ||
> 		    ((int32_t)req->value_us < new_max_latency_us)) {
> 			new_max_latency_us = (int32_t)req->value_us;
> 		}
> ```
> 
> Thread A registers request with latency of 2ms Thread B registers request with latency of SYS_FOREVER_US
> 
> 1. The snippet runs for the request from thread A, `new_max_latency_us` is set to 2000.
> 2. The snippet runs for the request from thread B, `new_max_latency_us` is 2000 however the request from B is `SYS_FOREVER_US`(-1). Since -1 < 2000, `new_max_latency_us` is now set to `SYS_FOREVER_US`(-1).
> 3. At the end of `update_max_latency()`, `max_latency_cyc` is set to -1
> 
> Now when `pm_system_suspend()` runs, `pm_policy_next_state()` will think there are no latency requests, and put the device in the lowest power mode possible depending on the ticks it can sleep for.

Yep, that is definitely a bug, the fix looks correct to me. There are problem in the other commit, fix them and we are good have this pr in.

siddharthgpta committed:
pm: policy: add a timeout for next event

The next event implementation is buggy. This is mainly due to the fact that
the next event information is not propagated to the kernel timeout manager.
This results in the kernel sleeping past the next event registered with the
pm policy.

Consider the following scenario:
1. The timeout manager sees the kernel can sleep for 10ms.
2. Thread A registers an event in 5ms.
3. The low power mode min residency is 4ms and exit latency is 500µs.

In this scenario:
- pm_system_suspend() will call pm_policy_next_state() with ticks
  equivalent to 10ms.
- pm_policy_next_state() will update cyc to cyc_evt since the next event
  is in 5ms.
- cyc is greater than the min residency + exit latency so sleep is allowed,
  and return next state as sleep
- pm_system_suspend() will set a timeout as 10ms because it does not see
  the next event

This fixes this issue by adding a timeout when the next event is due thus
allowing the timeout manager to provide an appropriate time for the device
to sleep.

Signed-off-by: Siddharth Gupta <siddharth.gupta2694@gmail.com>

siddharthgpta commented:
Hey @ceolin, I removed the second commit which was adding the next event subscription list in favour of propagating the next event through the kernel timer. I have added the information in the commit message. Please let me know if the commit message isn't clear enough!

ceolin commented:
> Hey @ceolin, I removed the second commit which was adding the next event subscription list in favour of propagating the next event through the kernel timeout manager. I have added the information in the commit message. Please let me know if the commit message isn't clear enough!

That is not correct, this changes the purpose of the event. The event is about something t that *will* happen and will wake up the system. We use it to select a state that will be able respond that event in a timely manner. 

What you are doing here is creating a new interruption re-programming the clock to wake the system in addition to the original event that will happen. Lets put this way, you application know that it will receive from serial every 10ms you use this event api to tell the policy not choose any power state that would cause problem to respond this interruption properly. The serial interruption will wake up the sysetem !

siddharthgpta commented:
> > Hey @ceolin, I removed the second commit which was adding the next event subscription list in favour of propagating the next event through the kernel timeout manager. I have added the information in the commit message. Please let me know if the commit message isn't clear enough!
> 
> That is not correct, this changes the purpose of the event. The event is about something t that _will_ happen and will wake up the system. We use it to select a state that will be able respond that event in a timely manner.
> 
> What you are doing here is creating a new interruption re-programming the clock to wake the system in addition to the original event that will happen. Lets put this way, you application know that it will receive from serial every 10ms you use this event api to tell the policy not choose any power state that would cause problem to respond this interruption properly. The serial interruption will wake up the sysetem !

Thanks for clarifying the expected usage!

So the problem I described in my commit message needs to be solved by utilizing the latency requests along with the next event information instead of relying on a timeout to force a wakeup?

ceolin commented:
> > > Hey @ceolin, I removed the second commit which was adding the next event subscription list in favour of propagating the next event through the kernel timeout manager. I have added the information in the commit message. Please let me know if the commit message isn't clear enough!
> > 
> > 
> > That is not correct, this changes the purpose of the event. The event is about something t that _will_ happen and will wake up the system. We use it to select a state that will be able respond that event in a timely manner.
> > What you are doing here is creating a new interruption re-programming the clock to wake the system in addition to the original event that will happen. Lets put this way, you application know that it will receive from serial every 10ms you use this event api to tell the policy not choose any power state that would cause problem to respond this interruption properly. The serial interruption will wake up the sysetem !
> 
> Thanks for clarifying the expected usage!
> 
> So the problem I described in my commit message needs to be solved by utilizing the latency requests along with the next event information instead of relying on a timeout to force a wakeup?

I can't see the problem. You have this problem because you have assumed that the system will sleep over the event, but what happens is that event will happen and will wake up the system.  We use this API to inform that subsystem that this external event (that is not scheduled in Zephyr, otherwise it will reflect in the next ticks) will happen so the subsystem will not select a power state that had min residency greater than the time to this "unknown" event.

siddharthgpta commented:
> > > > Hey @ceolin, I removed the second commit which was adding the next event subscription list in favour of propagating the next event through the kernel timeout manager. I have added the information in the commit message. Please let me know if the commit message isn't clear enough!
> > > 
> > > 
> > > That is not correct, this changes the purpose of the event. The event is about something t that _will_ happen and will wake up the system. We use it to select a state that will be able respond that event in a timely manner.
> > > What you are doing here is creating a new interruption re-programming the clock to wake the system in addition to the original event that will happen. Lets put this way, you application know that it will receive from serial every 10ms you use this event api to tell the policy not choose any power state that would cause problem to respond this interruption properly. The serial interruption will wake up the sysetem !
> > 
> > 
> > Thanks for clarifying the expected usage!
> > So the problem I described in my commit message needs to be solved by utilizing the latency requests along with the next event information instead of relying on a timeout to force a wakeup?
> 
> I can't see the problem. You have this problem because you have assumed that the system will sleep over the event, but what happens is that event will happen and will wake up the system. We use this API to inform that subsystem that this external event (that is not scheduled in Zephyr, otherwise it will reflect in the next ticks) will happen so the subsystem will not select a power state that had min residency greater than the time to this "unknown" event.

The problem is that the latency introduced by entering the sleep state is too high for the system to be able to respond to it in a timely manner. The only way to deal with it is to also specify a latency requirement from the thread to prevent the cpu from entering sleep state.
